sklearn/mixture/base.py

```diff
@@ -0,0 +1,441 @@
+                prev_log_likelihood = current_log_likelihood
+
+                current_log_likelihood, resp = self._e_step(X)
+                change = current_log_likelihood - prev_log_likelihood
+                self._print_verbose_msg_iter_end(n_iter, change)
+
+                if abs(change) < self.tol:
+                    self.converged_ = True
+                    break
+
+                self._m_step(X, resp)
+
+            # print verbose msg
+            self._print_verbose_msg_init_end(current_log_likelihood)
+
+            if current_log_likelihood > max_log_likelihood:
+                # max_log_likelihood is always updated,
+                # since we compute the log-likelihood of the initialization
+                max_log_likelihood = current_log_likelihood
+                best_params = self._get_parameters()
+                best_n_iter = n_iter
+
+        if not self.converged_:
+            warnings.warn('Initialization %d did not converged. '
+                          'Try different init parameters, '
+                          'or increase n_init, tol '
+                          'or check for degenerate data.'
+                          % (init + 1), ConvergenceWarning)
+            # compute the log-likelihood of the last m-step
+            current_log_likelihood, _ = self._e_step(X)
+
+        self._set_parameters(best_params)
+        self.n_iter_ = best_n_iter
+
+        return self
+
+    @abstractmethod
+    def _e_step(self, X):
+        """E step.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+
+        Returns
+        -------
+        log-likelihood : scalar
+
+        responsibility : array, shape (n_samples, n_components)
+        """
+        pass
+
+    @abstractmethod
+    def _m_step(self, X, resp):
+        """M step.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+
+        resp : array-like, shape (n_samples, n_components)
+        """
+        pass
+
+    @abstractmethod
+    def _check_is_fitted(self):
+        pass
+
+    @abstractmethod
+    def _get_parameters(self):
+        pass
+
+    @abstractmethod
+    def _set_parameters(self, params):
+        pass
+
+    def score_samples(self, X):
+        """Compute the weighted log probabilities for each sample.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            List of n_features-dimensional data points. Each row
+            corresponds to a single data point.
+
+        Returns
+        -------
+        log_prob : array, shape (n_samples,)
+            Log probabilities of each data point in X.
+        """
+        self._check_is_fitted()
+        X = _check_X(X, None, self.means_.shape[1])
+
+        return logsumexp(self._estimate_weighted_log_prob(X), axis=1)
+
+    def score(self, X, y=None):
+        """Compute the log-likelihood of the model on the given data X.
```

Comment:
The docstring is not correct: this is the per-sample average log-likelihood: as we use `p.mean()` instead of `p.sum()`.